### ISSUE TITLE
Phase 2: Temperature STE + EMA shadow weights + bfloat16 (+15-25% speedup)

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/trinity#294]
## Goal
Phase 2 training optimizations after Phase 1 (#293). Cumulative +15-25% speedup.

### 1. Temperature-Annealed STE Quantization
- Start with tau=5.0, anneal to tau=0.1 over training
- Soft ternary during forward, hard ternary during inference
- Expected: 10-15% convergence improvement

### 2. EMA Shadow Weights
- Maintain exponential moving average of weights (decay=0.999)
- Use EMA weights for evaluation/inference
- Expected: 5-8% PPL improvement

### 3. bfloat16 Shadow Weights
- Keep shadow copies in bfloat16 for gradient accumulation
- Quantize to ternary after update step
- Reduces memory, enables larger models

### 4. RMSNorm Before Linear Layers
- From BitNet b1.58: critical for ternary network stability
- Add RMSNorm before every linear projection

## Depends on
- #293 (Phase 1)

## Priority
P1 - days 3-5

### Files to modify

- src/tri-api/ (STE, EMA, optimizer)
- src/vsa.zig (quantization, RMSNorm)

### Verification steps

1. `zig build test` passes
2. Temperature annealing shows smooth loss curve
3. EMA weights give lower eval PPL than raw weights
4. RMSNorm stabilizes training at larger learning rates

**Comment by gHashTag** [comment]
🌅 **Trinity Agent Dashboard** — Issue #294

| Field | Value |
|-------|-------|
| **Status** | AWAKENING |
| **Step** | 1/8 |
| **Detail** | Authenticating with GitHub |
| **Elapsed** | 0s |
| **Container** | agent-294 |
| **Updated** | 2026-03-12T05:56:45Z |

**Comment by gHashTag** [comment]
🌅 **Trinity Agent Dashboard** — Issue #294

| Field | Value |
|-------|-------|
| **Status** | AWAKENING |
| **Step** | 1/8 |
| **Detail** | Authenticating with GitHub |
| **Elapsed** | 0s |
| **Container** | agent-294 |
| **Updated** | 2026-03-12T05:56:46Z |

**Comment by gHashTag** [comment]
⚡ **Trinity Agent Dashboard** — Issue #294

| Field | Value |
|-------|-------|
| **Status** | CODING |
| **Step** | 4/8 |
| **Detail** | Claude Code running (timeout: 3600s) |
| **Elapsed** | 12s |
| **Container** | agent-294 |
| **Updated** | 2026-03-12T05:56:57Z |

**Comment by gHashTag** [comment]
⚡ **Trinity Agent** | 2026-03-12T05:56:57Z
📋 **Step**: 4/8 — Claude Code running (timeout: 3600s)
🔄 **Status**: CODING
⏱️ **Elapsed**: 12s

**Comment by gHashTag** [comment]
🌅 **Trinity Agent** | 2026-03-12T05:57:15Z
📋 **Step**: 1/8 — Authenticating with GitHub
🔄 **Status**: AWAKENING
⏱️ **Elapsed**: 0s

**Comment by gHashTag** [comment]
🌅 **Trinity Agent Dashboard** — Issue #294

| Field | Value |
|-------|-------|
| **Status** | AWAKENING |
| **Step** | 1/8 |
| **Detail** | Authenticating with GitHub |
| **Elapsed** | 0s |
| **Container** | agent-294 |
| **Updated** | 2026-03-12T05:57:15Z |

**Comment by gHashTag** [comment]
⚡ **Trinity Agent Dashboard** — Issue #294

| Field | Value |
|-------|-------|
| **Status** | CODING |
| **Step** | 4/8 |
| **Detail** | Claude Code running (timeout: 3600s) |
| **Elapsed** | 16s |
| **Container** | agent-294 |
| **Updated** | 2026-03-12T05:57:31Z |

**Comment by gHashTag** [comment]
❌ **Trinity Agent**: Compilation gate FAILED after 3 repair attempts. Code does not build.

```
timeout: failed to run command 'zig': No such file or directory
```

PR creation skipped. Issue needs manual attention.

**Comment by gHashTag** [comment]
🔍 **Trinity Agent** | 2026-03-12T21:23:40Z
📋 **Step**: 5/8 — Self-review (advisory)
🔄 **Status**: REVIEWING
⏱️ **Elapsed**: 1186s

**Comment by gHashTag** [comment]
🔧 **Trinity Agent** | 2026-03-12T21:23:43Z
📋 **Step**: 6/8 — Compilation failed, auto-repair attempt 1/3
🔄 **Status**: REPAIRING
⏱️ **Elapsed**: 1189s

**Comment by gHashTag** [comment]
📖 **Trinity Agent** | 2026-03-12T21:27:59Z
📋 **Step**: 3/8 — Reading issue #294
🔄 **Status**: READING
⏱️ **Elapsed**: 10s

**Comment by gHashTag** [comment]
🔍 **Trinity Agent** | 2026-03-12T21:41:03Z
📋 **Step**: 6/8 — Self-review (advisory)
🔄 **Status**: REVIEWING
⏱️ **Elapsed**: 794s

**Comment by gHashTag** [comment]
🧪 **Trinity Agent** | 2026-03-12T22:12:55Z
📋 **Step**: 7/8 — verification passes. and `zig build test` passes, and `zig fmt src/ && zig build`, to see the change
🔄 **Status**: TESTING
⏱️ **Elapsed**: 743s

**Comment by gHashTag** [comment]
📖 **Trinity Agent** | 2026-03-12T22:13:38Z
📋 **Step**: 3/8 — Reading issue #294
🔄 **Status**: READING
⏱️ **Elapsed**: 9s
